### PR TITLE
UCP: Recv msg nbx routine.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
NBX routine for tag_recv_msg

## Why ?
Want to switch JUCX completely to NBX API. This routine is missing in implementation.

### CODE REVIEW

**Review by dmitrygx — CHANGES_REQUESTED** [review]

File: test/gtest/ucp/test_ucp_tag_probe.cc [hunk 1]
```diff
@@ -166,9 +166,15 @@ UCS_TEST_P(test_ucp_tag_probe, send_rndv_msg_probe, "RNDV_THRESH=1048576") {
     EXPECT_EQ((ucp_tag_t)0x111337, info.sender_tag);
 
     /* receiver - process the rts and schedule a get operation */
-    my_recv_req = (request*)ucp_tag_msg_recv_nb(receiver().worker(), &recvbuf[0],
-                                                recvbuf.size(), DATATYPE, message,
-                                                recv_callback);
+    ucp_request_param_t param;
+    param.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK | UCP_OP_ATTR_FIELD_DATATYPE |
+                         UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+    param.user_data    = NULL;
```
[the remark this answers is not in the document]
Comment (petro-rudenko) [reply test/gtest/ucp/test_ucp_tag_probe.cc:172]: There's some strange logic in https://github.com/openucx/ucx/blob/master/test/gtest/ucp/test_ucp_tag.cc#L110 
```
struct request *req = (struct request *)(user_data ?: request);
```
So it'll take `user_data` as request if it's not null.

File: src/ucp/tag/tag_recv.c [hunk 2]
```diff
@@ -252,3 +252,32 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_msg_recv_nb,
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
     return ret;
 }
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_msg_recv_nbx,
```
[the remark this answers is not in the document]
Comment (petro-rudenko) [reply src/ucp/tag/tag_recv.c:256]: It doesn't return a request, so for JUCX don't need it, maybe in other places.

[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/tag/tag_recv.c:256]: > It doesn't return a request, so for JUCX don't need it, maybe in other places.

I see

File: test/gtest/ucp/test_ucp_tag_probe.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_probe.cc:172]: hmm, it strange, but UCP expects that user_data is initialized, if `UCP_OP_ATTR_FIELD_CALLBACK` is set
https://github.com/openucx/ucx/blob/73a7f368a3488370d76caf3b13967d9bf6000622/src/ucp/tag/tag_recv.c#L102

File: src/ucp/tag/tag_recv.c [hunk 3]
```diff
@@ -221,36 +221,19 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_recv_nbx,
...
-                 ucp_tag_recv_callback_t cb)
+ucs_status_ptr_t ucp_tag_msg_recv_nb(ucp_worker_h worker, void *buffer, size_t count,
+                                     ucp_datatype_t datatype, ucp_tag_message_h message,
+                                     ucp_tag_recv_callback_t cb)
 {
     ucp_request_param_t param = {
         .op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                        UCP_OP_ATTR_FIELD_DATATYPE |
                         UCP_OP_ATTR_FLAG_NO_IMM_CMPL,
+        .datatype     = datatype,
```
Comment (dmitrygx) [inline src/ucp/tag/tag_recv.c:232]: [q] should we initialize `user_data = NULL` here like we did in the test?

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/tag/tag_recv.c (hunk 3 above)
Comment (petro-rudenko) [reply src/ucp/tag/tag_recv.c:232]: No, because in test there's a custom generic callback  https://github.com/openucx/ucx/blob/master/test/gtest/ucp/test_ucp_tag.cc#L110 - if a user will pass callback where he expects to get valid user_data but he didn't pass - it's a user error. However there's probably bug here:
https://github.com/openucx/ucx/blob/master/src/ucp/tag/tag_recv.c#L104

Since it sets user data, but checking only callback flag. WDYT?

Comment (dmitrygx) [reply src/ucp/tag/tag_recv.c:232]: > Since it sets user data, but checking only callback flag. WDYT?

yes, I think this is a bug. So, we also should fix the test to pass USER_DATA, even if it is `NULL`, I guess

File: src/ucp/tag/tag_recv.c [hunk 4]
```diff
@@ -101,7 +101,12 @@ ucp_tag_recv_common(ucp_worker_h worker, void *buffer, size_t count,
     req->recv.tag.tag_mask  = tag_mask;
     if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK) {
         req->recv.tag.cb    = param->cb.recv;
-        req->user_data      = param->user_data;
+
+        if (param->op_attr_mask & UCP_OP_ATTR_FIELD_USER_DATA) {
```
Comment (dmitrygx) [inline src/ucp/tag/tag_recv.c:105]: maybe move this assignment outside of `if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK)` statement?

Comment (petro-rudenko) [reply src/ucp/tag/tag_recv.c:105]: But user_data is needed / used only if callback is set.

Comment (dmitrygx) [reply src/ucp/tag/tag_recv.c:105]: ah, I see

**Comment by petro-rudenko** [comment]
bot:pipe:retest

File: src/ucp/tag/tag_recv.c [hunk 5]
```diff
@@ -221,34 +226,46 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_recv_nbx,
...
-                                  ucp_rdesc_get_tag(rdesc), UCP_TAG_MASK_FULL,
-                                  req, rdesc, &param, "msg_recv_nb");
-    } else {
-        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-    }
+    req = ucp_request_get_param(worker, param,
+                                {ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+                                goto out;});
 
+    datatype = ucp_request_param_datatype(param);
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c:262]: indent = on column

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@tonycurtis @shamisp can you pls review?

**Review by tonycurtis — CHANGES_REQUESTED** [review]
ucp.h:3502: The @ref....
3512: Pointer to the buffer that will receive the data

**Comment by shamisp** [comment]
@petro-rudenko please update news. Otherwise 👍

**Comment by petro-rudenko** [comment]
Squashed
